### PR TITLE
Remove Eager Trait for Joins

### PR DESCRIPTION
**Pull request by berkaysynnada**
## Which issue does this PR close?

Closes #.

## Rationale for this change

`EagerJoinStream` trait is currently implemented only by `SymmetricHashJoin`. It was initially designed to be open-ended for other joins that could potentially work with an eager polling strategy. However, this strategy does not provide significant benefits or serve as a better alternative to other strategies. It is specifically suited to the nature of `SymmetricHashJoin`. Therefore, the implementation has been moved directly to `SymmetricHashJoin`.

Although it was suggested that `SortMergeJoin` might utilize this strategy, it does not require such a polling mechanism.

For simplification, this PR removes the redundant trait implementation and moves the existing implementation to where it is actually needed. If this trait is being utilized downstream, please share your thoughts.

## What changes are included in this PR?

Removed `EagerJoinStream` and moved its implementation to `SymmetricHashJoin`.
Removed the `handle_async_state` macro, which does not simplify any complexity.

## Are there any user-facing changes?

There are no user-facing changes. This PR involves internal refactoring for simplification and does not impact public APIs.

### CODE REVIEW

**Review by ozankabak — APPROVED** [review]
LGTM, thank you for this simplification.

**Review by metegenez — APPROVED** [review]
LGTM as well.

**Comment by alamb** [comment]
cc @viirya  and @comphead  who I think are working on Sort Merge Join

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
Looks good to me. I don't see that SortMergeJoin is required to use this trait.

File: datafusion/core/src/datasource/listing/helpers.rs [hunk 1]
```diff
@@ -786,7 +786,7 @@ mod tests {
         assert_eq!(
             evaluate_partition_prefix(
                 partitions,
-                &[col("a").eq(lit("foo")).and((col("b").eq(lit("bar"))))],
+                &[col("a").eq(lit("foo")).and(col("b").eq(lit("bar")))],
```
Comment (comphead) [inline datafusion/core/src/datasource/listing/helpers.rs:789]: this already fixed. twice :)

**Review by comphead — APPROVED** [review]
lgtm

**Pull request merged**